### PR TITLE
RHSTOR 8248 - Add staggered CronJob scheduling edge-case tests

### PR DESCRIPTION
**Pull request by AviadP**
Test case Added:
- [x] invalid ConfigMap value
- [x] controller restart
- [x] short interval window shrink
- [x] UID-based offset determinism
- [x] missed run bypass
- [x] offset persistence across restart
- [x] thundering herd prevention with 10 PVCs
    
Make restart_csi_addons_controller public and add restart parameter to `update_csi_addons_config` 
Add helper methods for multi-round job collection, stagger offset calculation with circular distance, and direct CronJob suspend/resume

## Summary by CodeRabbit

* **New Features**
  * Improved controller restart behavior with smarter namespace resolution and an option to skip automatic restart after config updates.

* **Tests**
  * Expanded cronjob scheduling tests: stagger/window validation, timing tolerances, suspend/resume behavior, restart effects, offset consistency, invalid-config blocking, custom windows, and large-scale distribution.

* **Bug Fixes**
  * StorageClass patch now handles transient “not found” cases gracefully, logging a warning instead of failing.

### CODE REVIEW

File: ocs_ci/ocs/resources/csi_addons.py [hunk 1]
```diff
@@ -50,17 +53,21 @@ def _restart_csi_addons_controller(namespace: str) -> None:
             break
 
 
-def update_csi_addons_config(key: str, value: str) -> None:
+def update_csi_addons_config(key: str, value: str, restart: bool = True) -> None:
```
[the remark this answers is not in the document]
Comment (AviadP) [reply ocs_ci/ocs/resources/csi_addons.py]: Fixed. Added `*` to make `restart` keyword-only.

File: tests/functional/pv/space_reclaim/test_staggered_cronjob_scheduling.py [hunk 2]
```diff
@@ -302,6 +322,9 @@ def _collect_job_timestamps(
                 for cj_name, ts in timestamps.items():
                     logger.info("Job timestamp for CronJob '%s': %s", cj_name, ts)
                 return timestamps
+        raise TimeoutExpiredError(
+            f"Jobs not found for all CronJobs within {JOB_CREATION_TIMEOUT}s"
+        )
```
[the remark this answers is not in the document]
Comment (AviadP) [reply tests/functional/pv/space_reclaim/test_staggered_cronjob_scheduling.py]: Fixed. Error message now uses the `timeout` parameter instead of the hardcoded constant.

File: tests/functional/pv/space_reclaim/test_staggered_cronjob_scheduling.py [hunk 3]
```diff
@@ -328,3 +354,787 @@ def _assert_stagger_spread(self, timestamps: dict) -> None:
...
+        jobs_before_count = len(
+            [
+                j
+                for j in jobs_before
+                if j["metadata"]["name"].startswith(f"{cj_obj.resource_name}-")
+            ]
+        )
+
+        logger.info("Resuming CronJob '%s'", cj_obj.resource_name)
+        resume_time = datetime.utcnow()
```
[the remark this answers is not in the document]
Comment (AviadP) [reply tests/functional/pv/space_reclaim/test_staggered_cronjob_scheduling.py]: Fixed. Replaced with `datetime.now(timezone.utc).replace(tzinfo=None)`.

File: tests/functional/pv/space_reclaim/test_staggered_cronjob_scheduling.py [hunk 4]
```diff
@@ -328,3 +354,787 @@ def _assert_stagger_spread(self, timestamps: dict) -> None:
...
+        )
+
+        uids = set()
+        for pvc_name, cj_obj in cronjob_map.items():
+            uid = cj_obj.get()["metadata"]["uid"]
+            uids.add(uid)
+        assert len(uids) == NUM_PVCS_LARGE, (
+            f"Expected {NUM_PVCS_LARGE} unique CronJob UIDs, "
+            f"got {len(uids)}: {uids}"
+        )
```
[the remark this answers is not in the document]
Comment (AviadP) [reply tests/functional/pv/space_reclaim/test_staggered_cronjob_scheduling.py:1109]: Fixed. Changed to `.values()`.

**Comment by AviadP** [comment]
Regarding the outside-diff comment on shell quoting in `oc patch` (lines 90-93, 170-173): this is pre-existing code, not introduced by this PR. Current callers only pass simple strings (`"2"`, `"0"`, `"storageclass"`). Worth addressing separately.

File: tests/functional/pv/space_reclaim/test_staggered_cronjob_scheduling.py [hunk 5]
```diff
@@ -328,3 +350,785 @@ def _assert_stagger_spread(self, timestamps: dict) -> None:
...
+                delay = (job_time - resume_time).total_seconds()
+                logger.info(
+                    "Post-resume Job created %.1fs after resume",
+                    delay,
+                )
+                assert delay < JOB_IMMEDIATE_THRESHOLD_SECONDS, (
+                    f"Missed-run Job took {delay:.1f}s after resume, "
+                    f"expected within {JOB_IMMEDIATE_THRESHOLD_SECONDS}s"
+                )
+                break
```
[the remark this answers is not in the document]
Comment (AviadP) [reply tests/functional/pv/space_reclaim/test_staggered_cronjob_scheduling.py:964]: Not applicable — `TimeoutSampler.__iter__` is a `while True` loop that raises `TimeoutExpiredError` on timeout (see `utils.py:2052-2053`). It never terminates normally via `StopIteration`, so the `for-else` branch would be dead code.

File: tests/functional/pv/space_reclaim/test_staggered_cronjob_scheduling.py [hunk 6]
```diff
@@ -328,3 +350,785 @@ def _assert_stagger_spread(self, timestamps: dict) -> None:
...
+        )
+        found = cronjob_ocp.check_resource_existence(
+            should_exist=True, timeout=CRONJOB_CREATION_TIMEOUT
+        )
+        assert not found, (
+            f"CronJob '{expected_name}' was created despite invalid "
+            f"stagger window '{INVALID_STAGGER_WINDOW}'. If the controller "
+            f"now falls back to default, DFBUGS-6991 is fixed — update "
+            f"this test to verify fallback behavior."
+        )
```
[the remark this answers is not in the document]
Comment (AviadP) [reply tests/functional/pv/space_reclaim/test_staggered_cronjob_scheduling.py:717]: This tests a controller-level config rejection (invalid stagger window), not per-PVC behavior. If the controller rejects the config value, it blocks all CronJob creation — checking one PVC is sufficient to prove that. Iterating all 3 would triple the negative-wait time (3 × 180s) with no coverage gain.

File: tests/functional/pv/space_reclaim/test_staggered_cronjob_scheduling.py [hunk 7]
```diff
@@ -328,3 +350,785 @@ def _assert_stagger_spread(self, timestamps: dict) -> None:
...
+            pvc_namespace,
+            cronjob_names,
+            round_number=3,
+        )
+        ts_values = list(ts_round3.values())
+        max_spread = max(
+            abs((a - b).total_seconds())
+            for i, a in enumerate(ts_values)
+            for b in ts_values[i + 1 :]
+        )
```
Comment (yitzhak12) [inline tests/functional/pv/space_reclaim/test_staggered_cronjob_scheduling.py:788]: Better to have a comment or a log message when we have a complex calculation.

File: tests/functional/pv/space_reclaim/test_staggered_cronjob_scheduling.py [hunk 8]
```diff
@@ -328,3 +350,785 @@ def _assert_stagger_spread(self, timestamps: dict) -> None:
...
+                if j["metadata"]["name"].startswith(f"{cj_obj.resource_name}-")
+            ]
+            if len(matching) > jobs_before_count:
+                newest = max(
+                    matching,
+                    key=lambda j: j["metadata"]["creationTimestamp"],
+                )
+                ts_str = newest["metadata"]["creationTimestamp"]
+                job_time = datetime.strptime(ts_str, "%Y-%m-%dT%H:%M:%SZ")
+                delay = (job_time - resume_time).total_seconds()
```
Comment (yitzhak12) [inline tests/functional/pv/space_reclaim/test_staggered_cronjob_scheduling.py:955]: It would be better to add a comment or a log message here that explains the lines.

File: tests/functional/pv/space_reclaim/test_staggered_cronjob_scheduling.py [hunk 9]
```diff
@@ -328,3 +350,785 @@ def _assert_stagger_spread(self, timestamps: dict) -> None:
...
+            pvc_namespace, cronjob_names, timeout=LARGE_SCALE_JOB_TIMEOUT
+        )
+        self._assert_stagger_spread(timestamps)
+
+        ts_sorted = sorted(timestamps.values())
+        clustered = 0
+        for i in range(len(ts_sorted) - 1):
+            if abs((ts_sorted[i + 1] - ts_sorted[i]).total_seconds()) < 2:
+                clustered += 1
+        max_clustered = NUM_PVCS_LARGE // 3
```
Comment (yitzhak12) [inline tests/functional/pv/space_reclaim/test_staggered_cronjob_scheduling.py:1124]: Same here. It's better to include a comment or a log message that explains the code.

**Review by yitzhak12 — APPROVED** [review]

**Review by ebenahar — APPROVED** [review]
